types/src/proof/mod.rs

```diff
@@ -486,6 +486,115 @@ pub fn verify_sparse_merkle_element(
+///                                    /          \
+///                                  /              \
+///                                /                  \
+///                              /                      \
+///                            /                          \
+///                          /                              \
+///                        /                                  \
+/// frozen_subtree_roots[0]                                    o
+///                    /   \                                  / \
+///                   /     \                                /   \
+///                  o       o                              o     siblings[0]
+///                 / \     / \                            / \
+///                o   o   o   o    frozen_subtree_roots[1]   sibling[1]
+///               / \ / \ / \ / \                       / \
+///               o o o o o o o o                       o o
+/// ```
+///
+/// This function has at least two possible use cases:
+///   1) Given an accumulator represented by all of its frozen subtrees, we want to compute the
+///      root hash by combining the subtrees with placeholder nodes. In this case the caller can
+///      use a list of placeholder nodes as the `siblings` parameter.
+///   2) To prove that an accumulator A can be obtained by appending some elements to another
+///      accumulator B, the proof can provide a list of siblings and we can combine B's frozen
+///      subtrees with these siblings.
+pub fn get_accumulator_root_hash_by_frozen_subtrees_and_siblings<H: CryptoHasher>(
+    frozen_subtree_roots: &[HashValue],
+    num_leaves: u64,
+    siblings: &[HashValue],
+) -> HashValue {
+    // Special cases when there is zero or one subtree.
+    match frozen_subtree_roots.len() {
+        0 => return *ACCUMULATOR_PLACEHOLDER_HASH,
+        1 => return frozen_subtree_roots[0],
+        _ => (),
+    }
+
+    // For accumulators that have more than one frozen subtrees, the rightmost/smallest frozen
+    // subtree will always be combined with a sibling on the right.
+    let mut subtree_iter = frozen_subtree_roots.iter().rev();
+    let mut sibling_iter = siblings.iter().rev();
+    let mut current_hash = MerkleTreeInternalNode::<H>::new(
+        *subtree_iter.next().expect("Missing subtree hash."),
+        *sibling_iter.next().expect("Missing sibling hash."),
+    )
+    .hash();
+
+    // Remove the trailing zeros since they do not matter. Also remove one more bit since we have
+    // processed the lowest level.
+    let mut bitmap = num_leaves >> (num_leaves.trailing_zeros() + 1);
+    // Using the above example, now the bitmap becomes 0b10 (was 0b1010 originally). We check each
+    // bit from right to left.  If a bit is 0, it means there is no existing subtree on this level,
+    // so we combine current hash with a sibling.  Otherwise we combine it with the corresponding
+    // subtree.
+    while bitmap > 0 {
```

Comment:
the siblings can potentially go beyond the root of the original tree (to form a new tree several levels higher than the old tree), right?